### PR TITLE
Add sonar token for authentication.

### PR DESCRIPTION
**Pull request by swordbeta**
Should add optional authentication for SonarQube through a generated token.

I'm having a hard time testing if this work as I keep getting 0%. Could you perhaps help me out by telling where these logs can be found? https://github.com/jenkinsci/github-pr-coverage-status-plugin/blob/edf2f6a2403b0bf786e729b004fda2c44eb090ba/src/main/java/com/github/terma/jenkins/githubprcoveragestatus/SonarMasterCoverageRepository.java#L45
I don't see any in the job log and neither the system log. I don't see "Getting coverage for ..." anywhere either.

### CODE REVIEW

File: src/main/java/com/github/terma/jenkins/githubprcoveragestatus/SonarMasterCoverageRepository.java [hunk 1]
```diff
@@ -27,10 +29,14 @@
...
     private PrintStream buildLog;
 
-    public SonarMasterCoverageRepository(String sonarUrl, PrintStream buildLog) {
+    public SonarMasterCoverageRepository(String sonarUrl, String sonarToken, PrintStream buildLog) {
         this.sonarUrl = sonarUrl;
         this.buildLog = buildLog;
         httpClient = new HttpClient();
+        httpClient.getState().setCredentials(
+            new AuthScope(null, 0, null, null),
+            new UsernamePasswordCredentials(sonarToken, "")
```
Comment (terma) [inline src/main/java/com/github/terma/jenkins/githubprcoveragestatus/SonarMasterCoverageRepository.java]: So now we are going to have two case with ```sonarToken``` and without. Please make sure that in case without ```sonarToken``` code will work as it was before.

File: src/main/resources/com/github/terma/jenkins/githubprcoveragestatus/Configuration/help-sonarToken.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<div>
+    Access token for Sonar instance
+</div>
```
Comment (terma) [inline src/main/resources/com/github/terma/jenkins/githubprcoveragestatus/Configuration/help-sonarToken.html]: Thx a lot for help page 👍  Could you please specify here that ```sonarToken``` is optional.Additionally you can provide link on Sonar doc like https://docs.sonarqube.org/display/SONAR/User+Token

File: src/test/java/com/github/terma/jenkins/githubprcoveragestatus/SonarMasterCoverageRepositoryTest.java [hunk 3]
```diff
@@ -73,6 +73,7 @@ public void should_get_zero_coverage_for_unknown_metric() throws IOException {
     private void givenCoverageRepository() {
         buildLogOutputStream = new ByteArrayOutputStream();
         sonarMasterCoverageRepository = new SonarMasterCoverageRepository("http://localhost:" + wireMockRule.port(),
+                "myToken",
                 new PrintStream(buildLogOutputStream, true));
```
Comment (terma) [inline src/test/java/com/github/terma/jenkins/githubprcoveragestatus/SonarMasterCoverageRepositoryTest.java]: Please add at least two test cases here, one to demonstrate that in case of ```sonarToken = null``` we don't include any additional to ```request``` and test case for non empty

**Comment by swordbeta** [comment]
@terma Would be happy to apply the requested changes, but do you have any clue why I'm not seeing any logs from the `SonarMasterCoverageRepository` class?

**Comment by terma** [comment]
@swordbeta that's good question. I'm going to check that today after work. If any update I'll replay you.

**Comment by emoshaya-cognito** [comment]
Sonar authentication is exactly what I'm looking for.. Any update on this? 👍

**Comment by terma** [comment]
@swordbeta checked, I can see exceptions generated by ```SonarMasterCoverageRepository.java``` in build console. 
Steps:
- Create ```Freestyle project```
- Add post build actions ```Publish coverage to GitHub```
- Setup incorrect ```Sonar endpoint URL``` and turn on ```Use Sonar for master coverage```
- Run jenkins locally ```mvn hpi:run``` 
- I can see: 
```
Started by user anonymous
Building in workspace /Users/terma/Projects/github-pr-coverage-status-plugin/work/jobs/t1/workspace
Getting coverage for null
Failed to get master coverage for null
Exception message 'com.github.terma.jenkins.githubprcoveragestatus.SonarMasterCoverageRepository$SonarProjectRetrievalException: failed to search for sonar project opa/api/projects/index?search=null - host parameter is null'
com.github.terma.jenkins.githubprcoveragestatus.SonarMasterCoverageRepository$SonarProjectRetrievalException: failed to search for sonar project opa/api/projects/index?search=null - host parameter is null
	at com.github.terma.jenkins.githubprcoveragestatus.SonarMasterCoverageRepository.getSonarProject(SonarMasterCoverageRepository.java:75)
	at com.github.terma.jenkins.githubprcoveragestatus.SonarMasterCoverageRepository.get(SonarMasterCoverageRepository.java:40)
	at com.github.terma.jenkins.githubprcoveragestatus.CompareCoverageAction.perform(CompareCoverageAction.java:68)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:74)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:770)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:734)
	at hudson.model.Build$BuildExecution.post2(Build.java:183)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:683)
	at hudson.model.Run.execute(Run.java:1770)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:89)
	at hudson.model.Executor.run(Executor.java:240)
Caused by: java.lang.IllegalArgumentException: host parameter is null
	at org.apache.commons.httpclient.HttpConnection.<init>(HttpConnection.java:206)
	at org.apache.commons.httpclient.HttpConnection.<init>(HttpConnection.java:155)
	at org.apache.commons.httpclient.SimpleHttpConnectionManager.getConnectionWithTimeout(SimpleHttpConnectionManager.java:175)
	at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:153)
```

**Comment by aakavalevich** [comment]
It would be nice to have authentication with login and passwords also. For example, in our case we have a permanent login and OAuth token as a password. Token in Sonar you need to regenerate periodically manually and generation of such OAuth token can be integrated into pipelines, so no manual intervention is needed.

**Comment by terma** [comment]
Will be covered in #18
